assets/src/edit-story/components/library/panes/media/media3p/api/media3pApiProvider.js

```diff
@@ -0,0 +1,141 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import PropTypes from 'prop-types';
+
+/**
+ * Internal dependencies
+ */
+import { createResource } from '../../../../../../app/media/utils';
+import { listMedia as apiListMedia } from './apiFetcher';
+import Context from './context';
+
+/**
+ * The supported providers.
+ *
+ * @enum {string}
+ */
+const Providers = {
+  UNSPLASH: 'unsplash',
+};
+
+/**
+ * Provider for the Media3P API. Delegates fetching the data to apiFetcher,
+ * but transforms the response into resources.
+ */
+function Media3pApiProvider({ children }) {
+  const MEDIA_PAGE_SIZE = 20;
+
+  function constructFilter(provider, searchTerm, mediaType) {
+    let tokens = [];
+    if (provider) {
+      tokens.push(`provider:${provider}`);
+    }
+    if (mediaType) {
+      tokens.push(`type:${mediaType}`);
+    }
+    if (searchTerm) {
+      tokens.push(searchTerm);
+    }
+    return tokens.join(' ');
+  }
+
+  function getUrls(m) {
+    if (m.type.toLowerCase() === 'image') {
+      return Object.fromEntries(
+        new Map(m.imageUrls.map((u) => [u.imageName, u.url]))
+      );
+    }
+    throw new Error('Invalid media type.');
+  }
+
+  function getFullAsset(m) {
+    if (m.type.toLowerCase() === 'image') {
+      return m.imageUrls.find((i) => i.imageName === 'full');
+    }
+    throw new Error('Invalid media type.');
+  }
+
+  /**
+   * Get media for the given parameters.
+   *
+   * @param {Object} obj - An object with the options.
+   * @param {string} obj.provider The provider to get the media from. Currently only 'unsplash' is supported.
+   * @param {?string} obj.searchTerm Optional search term to send, eg: 'cute cats'.
+   * @param {?string} obj.orderBy The desired ordering of the results. Defaults to 'relevance' in the API.
+   * @param {?string} obj.mediaType The media type of results to get. Currently ignored by the API as Unsplash only handles images.
+   * @param {?string} obj.pageToken An optional page token to provide, for pagination.
+   * @return {Promise<{nextPageToken: *, media: *}>} An object with the media resources and a next page token.
+   */
+  async function listMedia({
+    provider,
+    searchTerm,
+    orderBy,
+    mediaType,
+    pageToken,
+  }) {
+    if (provider.toLowerCase() !== Providers.UNSPLASH) {
+      throw new Error(`Unsupported providers: ${provider}`);
+    }
+    const response = await apiListMedia({
+      filter: constructFilter(provider, searchTerm, mediaType),
+      orderBy: orderBy,
+      pageSize: MEDIA_PAGE_SIZE,
+      pageToken: pageToken,
+    });
+    return {
+      media: response.media.map((m) => {
```

Comment:
[nit]: Can this be a separate function, to make it easier to follow the logic inside of listMedia?